ui/src/components/Logs/useLogVirtualizer.ts

```diff
@@ -0,0 +1,145 @@
+        return uniqueId;
+      },
+      [logs]
+    ),
+    /**
+     * overscan is the number of items to render above and below
+     * the visible window. More items = less flickering when
+     * scrolling, but more memory usage and initial load time.
+     * I tested it with around 2000 items and 40 was a good fit
+     * to have no flickering with pretty high scrolling speed.
+     */
+    overscan: 40,
+    onChange(instance) {
+      prevNumberOfLogs.current = numberOfLogs;
+      prevOldestLogId.current = logs?.[0]?.id ?? null;
+      /**
+       * when the last x loglines are visible in the list (x being
+       * loglinesThreashold), the user is considered to be at the
+       * bottom of the list.
+       */
+      if (!instance.range) return;
+      const { endIndex: lastVisibleIndex } = instance.range;
+      const loglinesThreashold = 5;
+      setScrolledToBottom(
+        lastVisibleIndex >= numberOfLogs - 1 - loglinesThreashold
+      );
+    },
+  });
+
+  const virtualItems = rowVirtualizer.getVirtualItems();
+
+  /**
+   * fetch the previous log page when a users scrolls to the top
+   */
+  useEffect(() => {
+    const [firstLogEntry] = virtualItems;
+    if (
+      firstLogEntry?.index === 0 &&
+      hasPreviousPage &&
+      !isFetchingPreviousPage &&
+      prevNumberOfLogs?.current === numberOfLogs
+    ) {
+      fetchPreviousPage();
+    }
+  }, [
+    fetchPreviousPage,
+    hasPreviousPage,
+    isFetchingPreviousPage,
+    numberOfLogs,
+    virtualItems,
+  ]);
+
+  /**
+   * when the user reached the bottom of the list we need to keep the scoll
+   * position to the very bottom to make sure new log entries are in the
+   * viewport.
+   */
+  useEffect(() => {
+    if (numberOfLogs > 0 && scrolledToBottom) {
+      rowVirtualizer.scrollToIndex(numberOfLogs - 1, { align: "end" });
+    }
+  }, [numberOfLogs, rowVirtualizer, scrolledToBottom]);
+
+  /**
+   * maintain the scroll position when a set of new logs is added to the top
+   * of the list
+   */
+  useEffect(() => {
+    if (
+      !scrolledToBottom &&
+      prevNumberOfLogs.current &&
+      prevNumberOfLogs.current !== numberOfLogs &&
+      // this will make sure the new received logs were added at the top
+      prevOldestLogId.current !== (logs?.[0]?.id ?? null)
+    ) {
+      /**
+       * To maintin the old scroll position we need to know how many logs
+       * were added to the top of the list.
+       *
+       * Example:
+       * The user was at a scrollOffset of 100 and then 200 new logs have
+       * been added to the top. We now need to add 200 times the height of
+       * a log entry (200 * defaultLogHeight) to that offset to stay at the
+       * same scroll position.
+       */
+      const { scrollOffset: currentOffset } = rowVirtualizer;
+      const numberOfNewLogs = numberOfLogs - prevNumberOfLogs.current;
+      const newOffset = currentOffset + numberOfNewLogs * defaultLogHeight;
```

Comment:
That's a good catch. Since we don't scroll horizontally, I just disabled them via CSS (da8b1618b96cbb643548fa53a5f0efddc842aab4). However, I was not able to quickly introduce the extra padding that we talked about yesterday. When I add some simple padding, the scrolling gets weird, and I would rather avoid any "risky changes" for now, since we would need to start testing all behaviors again. 

Scrollbars are weird on a Mac, there are this big scrollbars like in your screenshot and this floating/invisible ones. By default, a MacBook has the floating scrollbars, but as soon as you connect a mouse, they change to the big scrollbars. I hate that extra scrolling space so much, that I disable them in my settings. That's why I did not see that before :) 

![CleanShot 2024-03-21 at 09 56 47@2x](https://github.com/direktiv/direktiv/assets/121789579/f753674e-674a-4379-a729-ec6ec4b652ee)